go-controller/pkg/retry/obj_retry_helper.go

```diff
@@ -0,0 +1,147 @@
+	r.retryEntries.LockKey(key)
+	defer r.retryEntries.UnlockKey(key)
+	obj, exists := r.getRetryObj(key)
+	if exists && obj != nil {
+		return obj.newObj
+	}
+	return nil
+}
+
+func GetConfigFromRetryObj(key string, r *RetryFramework) interface{} {
+	r.retryEntries.LockKey(key)
+	defer r.retryEntries.UnlockKey(key)
+	obj, exists := r.getRetryObj(key)
+	if exists && obj != nil {
+		return obj.config
+	}
+	return nil
+}
+
+func SetFailedAttemptsCounterForTestingOnly(key string, val uint8, r *RetryFramework) {
+	r.DoWithLock(key, func(key string) {
+		entry, found := r.getRetryObj(key)
+		if found {
+			entry.failedAttempts = val
+		}
+	})
+}
+
+func SetRetryObjWithNoBackoff(key string, r *RetryFramework) {
+	r.DoWithLock(key, func(key string) {
+		entry, found := r.getRetryObj(key)
+		if found {
+			r.setRetryObjWithNoBackoff(entry)
+		}
+	})
+}
+
+func InitRetryObjWithAdd(obj interface{}, key string, r *RetryFramework) {
+	r.DoWithLock(key, func(key string) {
+		r.initRetryObjWithAdd(obj, key)
+	})
+}
+
+func DeleteRetryObj(key string, r *RetryFramework) {
+	r.DoWithLock(key, func(key string) {
+		r.DeleteRetryObj(key)
+	})
+}
+
+func CheckRetryObjectEventually(key string, shouldExist bool, r *RetryFramework) {
+	expectedValue := gomega.BeTrue()
+	if !shouldExist {
+		expectedValue = gomega.BeFalse()
+	}
+	gomega.Eventually(func() bool {
+		return CheckRetryObj(key, r)
+	}, inspectTimeout).Should(expectedValue)
+}
+
+// CheckRetryObjectMultipleFieldsEventually verifies that eventually the oldObj, newObj, config and
+// failedAttemptsfields fields all satisfy the input conditions expectedParams, given in
+// the same order. In order not to check any of these four fields, the corresponding input
+// type.GomegaMatcher must be set to nil.
+func CheckRetryObjectMultipleFieldsEventually(
+	key string,
+	r *RetryFramework,
+	expectedParams ...types.GomegaMatcher) {
+
+	var expectedNewObj, expectedOldObj, expectedConfig, expectedFailedAttempts types.GomegaMatcher
+	for i, expectedParam := range expectedParams {
+		if i == 0 {
+			expectedOldObj = expectedParam
+		} else if i == 1 {
+			expectedNewObj = expectedParam
+		} else if i == 2 {
+			expectedConfig = expectedParam
+		} else if i == 3 {
+			expectedFailedAttempts = expectedParam
+		}
+	}
+
+	gomega.Eventually(func(g gomega.Gomega) {
+		r.DoWithLock(key, func(key string) {
+			obj, exists := r.getRetryObj(key)
+			g.Expect(exists).To(gomega.BeTrue())
+			g.Expect(obj).NotTo(gomega.BeNil())
+
+			if exists {
+				if expectedOldObj != nil {
+					g.Expect(obj.oldObj).To(expectedOldObj)
+				}
+				if expectedNewObj != nil {
+					g.Expect(obj.newObj).To(expectedNewObj)
+				}
+				if expectedConfig != nil {
+					g.Expect(obj.config).To(expectedConfig)
+				}
+				if expectedFailedAttempts != nil {
+					g.Expect(obj.failedAttempts).To(expectedFailedAttempts)
+				}
+			}
+		})
+	}, inspectTimeout).Should(gomega.Succeed())
+}
```

Comment:
hmm not sure if its ok to include gomega functions in none _test file ? is this file only used by UT maybe we can name it _test.go ?

Reply:
The file was named initially retry_test.go, but it turns out that _test source files are not compiled (and thus not exported) when building _other_ packages: https://pkg.go.dev/cmd/go#hdr-Compile_packages_and_dependencies (`When compiling packages, build ignores files that end in '_test.go'.`). So with this file named retry_test.go, its functions can only be used inside this package and not in `pkg/ovn`. That's why I renamed it.

Reply:
And as for including gomega functions, it works :)